internal/ui/concurrency/process_download_concurrency_model.go

```diff
@@ -0,0 +1,274 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui_concurrency
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/charmbracelet/bubbles/progress"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+
+	"tidbcloud-cli/internal/util"
+)
+
+var helpStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("#626262")).Render
+
+const (
+	padding  = 2
+	maxWidth = 80
+)
+
+type ProgressMsg struct {
+	id      int
+	percent float64
+}
+
+func NewProgressMsg(id int, percent float64) ProgressMsg {
+	return ProgressMsg{id: id, percent: percent}
+}
+
+type ProgressErrMsg struct {
+	id  int
+	err error
+}
+
+func NewProgressErrMsg(id int, err error) ProgressErrMsg {
+	return ProgressErrMsg{id: id, err: err}
+}
+
+type FileJob struct {
+	id      int
+	path    string
+	name    string
+	url     string
+	process progress.Model
+	err     error
+}
+
+type JobInfo struct {
+	idToJob       map[int]*FileJob
+	viewJobs      []*FileJob
+	finishedCount int
+	total         int
+	lock          sync.Mutex
+}
+
+type Model struct {
+	concurrency int
+	jobsCh      chan *FileJob
+	jobInfo     *JobInfo
+	width       int
+	once        sync.Once
+	onProgress  func(int, float64)
+	onError     func(int, error)
+	Interrupted bool
+}
+
+type URLMsg struct {
+	Name string
+	Path string
+	Url  string
+}
+
+func NewModel(urls []URLMsg, onProgress func(int, float64), onError func(int, error), concurrency int) Model {
+	jobs := make(chan *FileJob, len(urls))
+	idToJob := make(map[int]*FileJob)
+
+	for i, url := range urls {
+		job := &FileJob{id: i, name: url.Name, url: url.Url, path: url.Path}
+		idToJob[i] = job
+	}
+
+	jobInfo := &JobInfo{
+		idToJob:       idToJob,
+		viewJobs:      make([]*FileJob, 0, len(urls)),
+		finishedCount: 0,
+		total:         len(urls),
+	}
+	return Model{
+		jobsCh:      jobs,
+		jobInfo:     jobInfo,
+		concurrency: concurrency,
+		onProgress:  onProgress,
+		onError:     onError,
+	}
+}
+
+func NewDefaultModel(urls []URLMsg, onProgress func(int, float64), onError func(int, error)) Model {
+	return NewModel(urls, onProgress, onError, 2)
+}
+
+func (m Model) Init() tea.Cmd {
+	return nil
+}
+
+func (m *Model) InitPool() {
+	m.once.Do(func() {
+		// start consumer goroutine
+		for i := 0; i < m.concurrency; i++ {
+			go m.consume(m.jobsCh)
+		}
+		// start produce
+		for _, job := range m.jobInfo.idToJob {
+			go m.produce(job, m.jobsCh)
+		}
+	})
+}
+
+func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		if msg.Type == tea.KeyEsc || msg.Type == tea.KeyCtrlC {
+			m.Interrupted = true
+			return m, tea.Quit
+		}
+		return m, nil
+
+	case tea.WindowSizeMsg:
+		m.width = msg.Width
+		m.InitPool()
+
+		return m, nil
```

Comment:
https://pkg.go.dev/github.com/charmbracelet/bubbletea#WindowSizeMsg Why terminal resize need to initpool everytime?